cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -207,48 +258,53 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?>
     if (!schema.isIndexColumn(filterIndex.getName())) {
       throw new InvalidFieldException(schema.getTableId(), filterIndex.getName(), "is not an indexed column");
     }
-    ScannerIterator scannerIterator = new ScannerIterator(getScanner(keyRange), schema);
-    FilterByIndexIterator filterByIndexIterator = new FilterByIndexIterator(scannerIterator, filterIndex, schema);
+    FilterByFieldIterator filterByIndexIterator = new FilterByFieldIterator(getFilterByRangeIterator(keyRange),
+                                                                            filterIndex, schema);
     return new LimitIterator(Collections.singleton(filterByIndexIterator).iterator(), limit);
   }
 
+  /*
+   * Scan the nosql DB with provided ranges, then we de-duplicate the elements
+   * and sort the result in primary keys order in memory.
+   * */
   @Override
   public CloseableIterator<StructuredRow> multiScan(Collection<Range> keyRanges,
                                                     int limit) throws InvalidFieldException, IOException {
-    // Sort the scan keys by the start key and merge overlapping ranges.
-    Deque<ImmutablePair<byte[], byte[]>> scanKeys = new LinkedList<>();
-    keyRanges.stream()
-      .map(this::createScanKeys)
-      .sorted((o1, o2) -> Bytes.compareTo(o1.getFirst(), o2.getFirst()))
-      .forEach(range -> {
-        if (scanKeys.isEmpty()) {
-          scanKeys.add(range);
-        } else {
-          ImmutablePair<byte[], byte[]> last = scanKeys.getLast();
-          if (Bytes.compareTo(last.getSecond(), range.getFirst()) < 0) {
-            // No overlap
-            scanKeys.add(range);
-          } else {
-            // Combine overlapping ranges
-            scanKeys.pollLast();
-            byte[] end = Bytes.compareTo(last.getSecond(), range.getSecond()) > 0
-              ? last.getSecond() : range.getSecond();
-            scanKeys.addLast(ImmutablePair.of(last.getFirst(), end));
+    // TODO: CDAP-19688, instead of scanning all ranges, we could scan the most outbound range and do post filtering
```

Comment:
Opened PR here: https://github.com/cdapio/cdap/pull/14623

Didn't push to this PR since the build is taking so long